File: 2456-construct-smallest-number-from-di-string/construct-smallest-number-from-di-string.java
class Solution {
    public String smallestNumber(String pattern) {
        StringBuilder num = new StringBuilder();
        int cnt = 1, n = pattern.length();
        Stack<Character> st = new Stack<>();
        for(int i = 0; i<=n; i++){
            st.push((char)(cnt+'0'));
            cnt++;
            if(i<n && pattern.charAt(i) == 'I'){
                while(!st.isEmpty()){
                    num.append(st.pop());
                }
            }
        }
        while(!st.isEmpty()){
            num.append(st.pop());
        }
        return num.toString();
    }
}

File: 2456-construct-smallest-number-from-di-string/construct-smallest-number-from-di-string.cpp
class Solution {
public:
    string smallestNumber(string pattern) {
        string num;
        int cnt = 1, n = pattern.size();
        stack<char>st;
        for(int i = 0; i<=n; i++){
            if(pattern[i] == 'I'){
                st.push(cnt+'0');
                cnt++;
                while(!st.empty()){
                    num.push_back(st.top());
                    st.pop();
                }
            }
            else{
                st.push(cnt+'0');
                cnt++;
            }
        }
        while(!st.empty()){
            num.push_back(st.top());
            st.pop();
        }
        return num;
    }
};